File: src/Scanner.Core/Checks/AbsolutePositioningCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for absolute positioning that can block reflow.
/// </summary>
public sealed class AbsolutePositioningCheck : ICheck
{
    private static readonly Regex TagRegex = new("<(?<tag>[a-zA-Z0-9:-]+)(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly string[] CanvasAttributes =
    {
        "Canvas.Left",
        "Canvas.Top",
        "Canvas.Right",
        "Canvas.Bottom",
        "AbsoluteLayout.LayoutBounds"
    };

    /// <inheritdoc />
    public string Id => "absolute-positioning";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor", "xaml" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        if (context.Kind.Equals("xaml", StringComparison.OrdinalIgnoreCase))
        {
            return RunXaml(context, rule);
        }

        return RunMarkup(context, rule);
    }

    private static IEnumerable<Issue> RunMarkup(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in TagRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var style = AttributeParser.GetAttributeValue(attrs, "style");
            var position = StyleUtilities.GetLastPropertyValue(style, "position");
            if (!string.Equals(position, "absolute", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            yield return new Issue(rule.Id, Id, context.FilePath, line, "Element uses absolute positioning.", match.Value);
        }
    }

    private static IEnumerable<Issue> RunXaml(CheckContext context, RuleDefinition rule)
    {
        foreach (Match match in TagRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            if (!HasCanvasPositioning(attrs))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            yield return new Issue(rule.Id, Id, context.FilePath, line, "XAML element uses absolute positioning.", match.Value);
        }
    }

    private static bool HasCanvasPositioning(string attrs)
    {
        foreach (var attribute in CanvasAttributes)
        {
            var value = AttributeParser.GetAttributeValue(attrs, attribute);
            if (!string.IsNullOrWhiteSpace(value))
            {
                return true;
            }
        }

        return false;
    }
}

File: node_engine/src/checks/AbsolutePositioningCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber } = require("./TextUtilities");
const { getLastPropertyValue } = require("./StyleUtilities");

const tagRegex = /<(?<tag>[a-zA-Z0-9:-]+)(?<attrs>[^>]*)>/gi;
const canvasAttributes = [
  "Canvas.Left",
  "Canvas.Top",
  "Canvas.Right",
  "Canvas.Bottom",
  "AbsoluteLayout.LayoutBounds"
];
const positioningValues = ["absolute", "fixed"];

const isAbsolutePositioningValue = (value) =>
  Boolean(value && positioningValues.some((position) => position.toLowerCase() === value.toLowerCase()));

const tryGetCanvasPositioning = (attrs) => {
  for (const attribute of canvasAttributes) {
    const value = getAttributeValue(attrs, attribute);
    if (value && value.trim()) {
      return { attributeName: attribute, attributeValue: value };
    }
  }

  return null;
};

const runMarkup = (context, rule) => {
  const issues = [];
  for (const match of context.content.matchAll(tagRegex)) {
    const attrs = match.groups.attrs;
    const style = getAttributeValue(attrs, "style");
    const position = getLastPropertyValue(style, "position");
    if (!isAbsolutePositioningValue(position)) {
      continue;
    }

    const line = getLineNumber(context.content, match.index);
    issues.push({
      ruleId: rule.id,
      checkId: AbsolutePositioningCheck.id,
      filePath: context.filePath,
      line,
      message: `Element uses position: ${position}.`,
      evidence: match[0]
    });
  }

  return issues;
};

const runXaml = (context, rule) => {
  const issues = [];
  for (const match of context.content.matchAll(tagRegex)) {
    const attrs = match.groups.attrs;
    const canvas = tryGetCanvasPositioning(attrs);
    if (!canvas) {
      continue;
    }

    const line = getLineNumber(context.content, match.index);
    issues.push({
      ruleId: rule.id,
      checkId: AbsolutePositioningCheck.id,
      filePath: context.filePath,
      line,
      message: `XAML element uses absolute positioning (${canvas.attributeName}="${canvas.attributeValue}").`,
      evidence: match[0]
    });
  }

  return issues;
};

const AbsolutePositioningCheck = {
  id: "absolute-positioning",
  applicableKinds: ["html", "htm", "cshtml", "razor", "xaml"],
  run(context, rule) {
    if (context.kind.toLowerCase() === "xaml") {
      return runXaml(context, rule);
    }

    return runMarkup(context, rule);
  }
};

module.exports = { AbsolutePositioningCheck, isAbsolutePositioningValue, tryGetCanvasPositioning };
